File: permutationII/permutationII.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <memory>

using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        tmp = new int [nums.size()];
        sort(nums.begin(), nums.end()); 
        permuteUnique(nums, 0, res);
        delete [] tmp;
        return res;
    }
private:
	void permuteUnique(vector<int>& nums, int from, vector<vector<int>> &res) {
		if (from == nums.size()) {
			vector<int> t(tmp, tmp + from + 1);
			res.push_back(t);
			return;
		}
		for (int i = from; i < nums.size(); ++i) {
			tmp[i] = nums[i];
			vector<int> newnums(nums.begin(), nums.begin() + from);
			newnums
			permuteUnique(newnums, int from + 1, res)
		}
	}
	static int *tmp;
};

int *Solution::tmp = nullptr;

void compare(vector<vector<int>> &res, vector<vector<int>> &exp)
{
	if (res.size() != exp.size()) {
		cout << "different size" << endl;
	}
}

int main()
{
	Solution s;
	{
		vector<int> nums{1,1,2};
		vector<vector<int>> res = s.permuteUnique(nums);
		vector<vector<int>> exp{{1,1,2},{1,2,1},{2,1,1}};
		compare(res, exp);
	}
	{
		vector<int> nums{1,2,3};
		vector<vector<int>> res = s.permuteUnique(nums);
		vector<vector<int>> exp{{1,2,3},{1,3,2},{2,1,3},{2,3,1},{3,1,2},{3,2,1}};
		compare(res, exp);
	}
}

File: permutationII/permutationII.py
class Solution(object):
	def permuteUnique(self, nums):
		"""
		:type nums: List[int]
		:rtype: List[List[int]]
		"""
		ll = len(nums)
		if ll < 1:
			return [nums]
		res = []
		self.tmp = [0] * len(nums)
		nums.sort()
		self.permuteUnique2(nums, 0, res)
		return res

	def permuteUnique2(self, nums, curr, res):
		if len(nums) == 0:
			t = self.tmp[:]
			res.append(t)
			return

		prev = '$'
		for i, val in enumerate(nums):
			if nums[i] == prev:
				continue
			prev = nums[i]
			self.tmp[curr] = nums[i];
			newnums = nums[:i] + nums[i+1:]
			self.permuteUnique2(newnums, curr + 1, res)

		return


def compare(res, exp):
	if len(res) != len(exp):
		print("different size")
		return
	print("same size")

def main():
	s = Solution()

	nums = [1,1,2]
	res = s.permuteUnique(nums)
	exp = [[1,1,2],[1,2,1],[2,1,1]]
	compare(res, exp)
	print(res)
	print(exp)

	nums = [1,2,3]
	res = s.permuteUnique(nums)
	exp = [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
	compare(res, exp)
	print(res)
	print(exp)

if __name__ == "__main__":
     main()